File: infrastructures/ddd/csharp/src/Api/Controllers/ExampleController.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using OddlyDdd.Api.Dto.V1;
using OddlyDdd.Application.Services;

namespace OddlyDdd.Api.Controllers
{
    /// <summary>
    /// Example controller demonstrating controller layer patterns.
    /// Controllers:
    /// - Live in /Api/Controllers/
    /// - Have Controller suffix
    /// - Handle HTTP only: bind, validate, authorize, map DTOs
    /// - NO business logic (delegate to services)
    /// - Use BaseController for common functionality
    /// 
    /// Middleware order (automatically applied):
    /// 1. Correlation ID → 2. Logging → 3. Authentication → 
    /// 4. Authorization/Ownership → 5. UnitOfWork → 6. Controller →
    /// 7. UnitOfWork commit/rollback → 8. Error Handling
    /// </summary>
    [Authorize]
    public class ExampleController : BaseController
    {
        private readonly IExampleService _exampleService;

        public ExampleController(IExampleService p_exampleService)
        {
            _exampleService = p_exampleService;
        }

        /// <summary>
        /// Creates a new example.
        /// POST /api/v1/example
        /// </summary>
        [HttpPost]
        [ProducesResponseType(typeof(string), 201)]
        [ProducesResponseType(typeof(ErrorResponseDto), 400)]
        [ProducesResponseType(typeof(ErrorResponseDto), 401)]
        public async Task<IActionResult> CreateExample([FromBody] CreateExampleRequest p_request)
        {
            // Validate request (edge validation)
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            // Get user ID from auth context
            var userId = GetUserId();
            var correlationId = GetCorrelationId();

            // Delegate to service (business logic happens here)
            var exampleId = await _exampleService.CreateExampleAsync(p_request, userId, correlationId);

            // Return created response
            return CreatedAtAction(
                nameof(GetExample),
                new { id = exampleId },
                exampleId
            );
        }

        /// <summary>
        /// Gets an example by ID.
        /// GET /api/v1/example/{id}
        /// </summary>
        [HttpGet("{id}")]
        [ProducesResponseType(typeof(ExampleResponse), 200)]
        [ProducesResponseType(typeof(ErrorResponseDto), 404)]
        public async Task<IActionResult> GetExample(string id)
        {
            var response = await _exampleService.GetExampleAsync(id);
            return Ok(response);
        }

        /// <summary>
        /// Lists examples with pagination.
        /// GET /api/v1/example?skip=0&take=10
        /// </summary>
        [HttpGet]
        [ProducesResponseType(typeof(IReadOnlyList<ExampleResponse>), 200)]
        public async Task<IActionResult> ListExamples([FromQuery] int skip = 0, [FromQuery] int take = 10)
        {
            var responses = await _exampleService.ListExamplesAsync(skip, take);
            return Ok(responses);
        }

        /// <summary>
        /// Updates an example.
        /// PUT /api/v1/example/{id}
        /// </summary>
        [HttpPut("{id}")]
        [ProducesResponseType(204)]
        [ProducesResponseType(typeof(ErrorResponseDto), 400)]
        [ProducesResponseType(typeof(ErrorResponseDto), 403)]
        [ProducesResponseType(typeof(ErrorResponseDto), 404)]
        public async Task<IActionResult> UpdateExample(string id, [FromBody] UpdateExampleRequest p_request)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            var userId = GetUserId();
            await _exampleService.UpdateExampleAsync(id, p_request, userId);

            return NoContent();
        }

        /// <summary>
        /// Deletes an example.
        /// DELETE /api/v1/example/{id}
        /// </summary>
        [HttpDelete("{id}")]
        [ProducesResponseType(204)]
        [ProducesResponseType(typeof(ErrorResponseDto), 403)]
        [ProducesResponseType(typeof(ErrorResponseDto), 404)]
        public async Task<IActionResult> DeleteExample(string id)
        {
            var userId = GetUserId();
            await _exampleService.DeleteExampleAsync(id, userId);

            return NoContent();
        }

        /// <summary>
        /// Activates an example.
        /// POST /api/v1/example/{id}/activate
        /// </summary>
        [HttpPost("{id}/activate")]
        [ProducesResponseType(204)]
        [ProducesResponseType(typeof(ErrorResponseDto), 403)]
        [ProducesResponseType(typeof(ErrorResponseDto), 404)]
        public async Task<IActionResult> ActivateExample(string id)
        {
            var userId = GetUserId();
            await _exampleService.ActivateExampleAsync(id, userId);

            return NoContent();
        }

        /// <summary>
        /// Deactivates an example.
        /// POST /api/v1/example/{id}/deactivate
        /// </summary>
        [HttpPost("{id}/deactivate")]
        [ProducesResponseType(204)]
        [ProducesResponseType(typeof(ErrorResponseDto), 403)]
        [ProducesResponseType(typeof(ErrorResponseDto), 404)]
        public async Task<IActionResult> DeactivateExample(string id)
        {
            var userId = GetUserId();
            await _exampleService.DeactivateExampleAsync(id, userId);

            return NoContent();
        }
    }
}

File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/api/controllers/ExampleController.java
package com.oddly.ddd.api.controllers;

import com.oddly.ddd.api.controllers.infra.BaseController;
import com.oddly.ddd.api.dto.v1.requests.CreateExampleRequest;
import com.oddly.ddd.api.dto.v1.requests.UpdateExampleRequest;
import com.oddly.ddd.api.dto.v1.responses.ErrorResponseDto;
import com.oddly.ddd.api.dto.v1.responses.ExampleResponse;
import com.oddly.ddd.application.services.IExampleService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Example controller demonstrating controller layer patterns.
 * Controllers:
 * - Live in /api/controllers/
 * - Have Controller suffix
 * - Handle HTTP only: bind, validate, authorize, map DTOs
 * - NO business logic (delegate to services)
 * - Use BaseController for common functionality
 * 
 * Middleware order (automatically applied):
 * 1. Correlation ID → 2. Logging → 3. Authentication →
 * 4. Authorization/Ownership → 5. UnitOfWork → 6. Controller →
 * 7. UnitOfWork commit/rollback → 8. Error Handling
 */
@RestController
@RequestMapping("/api/v1/examples")
@PreAuthorize("isAuthenticated()")
public class ExampleController extends BaseController {
    private final IExampleService m_exampleService;

    public ExampleController(IExampleService p_exampleService) {
        this.m_exampleService = p_exampleService;
    }

    /**
     * Creates a new example.
     * POST /api/v1/examples
     */
    @PostMapping
    public CompletableFuture<ResponseEntity<String>> createExample(
            @Valid @RequestBody CreateExampleRequest p_request) {
        
        validateRequest(p_request);
        
        String userId = getUserId();
        String correlationId = getCorrelationId();
        
        return m_exampleService.createExampleAsync(p_request, userId, correlationId)
            .thenApply(id -> 
                ResponseEntity.created(URI.create("/api/v1/examples/" + id))
                    .body(id)
            );
    }

    /**
     * Gets an example by ID.
     * GET /api/v1/examples/{id}
     */
    @GetMapping("/{id}")
    public CompletableFuture<ResponseEntity<ExampleResponse>> getExample(
            @PathVariable String id) {
        
        return m_exampleService.getExampleAsync(id)
            .thenApply(ResponseEntity::ok);
    }

    /**
     * Lists examples with pagination.
     * GET /api/v1/examples?skip=0&take=10
     */
    @GetMapping
    public CompletableFuture<ResponseEntity<List<ExampleResponse>>> listExamples(
            @RequestParam(defaultValue = "0") int skip,
            @RequestParam(defaultValue = "10") int take) {
        
        return m_exampleService.listExamplesAsync(skip, take)
            .thenApply(ResponseEntity::ok);
    }

    /**
     * Updates an example.
     * PUT /api/v1/examples/{id}
     */
    @PutMapping("/{id}")
    public CompletableFuture<ResponseEntity<Void>> updateExample(
            @PathVariable String id,
            @Valid @RequestBody UpdateExampleRequest p_request) {
        
        validateRequest(p_request);
        
        String userId = getUserId();
        
        return m_exampleService.updateExampleAsync(id, p_request, userId)
            .thenApply(v -> ResponseEntity.noContent().build());
    }

    /**
     * Deletes an example.
     * DELETE /api/v1/examples/{id}
     */
    @DeleteMapping("/{id}")
    public CompletableFuture<ResponseEntity<Void>> deleteExample(
            @PathVariable String id) {
        
        String userId = getUserId();
        
        return m_exampleService.deleteExampleAsync(id, userId)
            .thenApply(v -> ResponseEntity.noContent().build());
    }

    /**
     * Activates an example.
     * POST /api/v1/examples/{id}/activate
     */
    @PostMapping("/{id}/activate")
    public CompletableFuture<ResponseEntity<Void>> activateExample(
            @PathVariable String id) {
        
        String userId = getUserId();
        
        return m_exampleService.activateExampleAsync(id, userId)
            .thenApply(v -> ResponseEntity.noContent().build());
    }

    /**
     * Deactivates an example.
     * POST /api/v1/examples/{id}/deactivate
     */
    @PostMapping("/{id}/deactivate")
    public CompletableFuture<ResponseEntity<Void>> deactivateExample(
            @PathVariable String id) {
        
        String userId = getUserId();
        
        return m_exampleService.deactivateExampleAsync(id, userId)
            .thenApply(v -> ResponseEntity.noContent().build());
    }
}
